src/java.base/share/classes/jdk/internal/misc/X-ScopedMemoryAccess.java.template

```diff
@@ -550,6 +550,191 @@ public class ScopedMemoryAccess {
+                                  int length, MemorySegmentProxy msp, long offset, M m, S s,
+                                  VectorSupport.LoadVectorMaskedOperation<MemorySegmentProxy, V, S, M> defaultImpl) {
+        // @@@ Smarter alignment checking if accessing heap segment backing non-byte[] array
+        if (msp.maxAlignMask() > 1) {
+            throw new IllegalArgumentException();
+        }
+
+        try {
+            return loadFromMemorySegmentMaskedScoped(
+                    msp.scope(),
+                    vmClass, maskClass, e, length,
+                    msp, offset, m,
+                    s,
+                    defaultImpl);
+        } catch (ScopedMemoryAccess.Scope.ScopedAccessError ex) {
+            throw new IllegalStateException("This segment is already closed");
+        }
+    }
+
+    @Scoped
+    @ForceInline
+    private static
+    <V extends VectorSupport.Vector<E>, E, S extends VectorSupport.VectorSpecies<E>,
+     M extends VectorSupport.VectorMask<E>>
+    V loadFromMemorySegmentMaskedScoped(ScopedMemoryAccess.Scope scope, Class<? extends V> vmClass,
+                                        Class<M> maskClass, Class<E> e, int length,
+                                        MemorySegmentProxy msp, long offset, M m,
+                                        S s,
+                                        VectorSupport.LoadVectorMaskedOperation<MemorySegmentProxy, V, S, M> defaultImpl) {
+        try {
+            scope.checkValidState();
+
+            return VectorSupport.loadMasked(vmClass, maskClass, e, length,
+                    msp.unsafeGetBase(), msp.unsafeGetOffset() + offset, m,
+                    msp, offset, s,
+                    defaultImpl);
+        } finally {
+            Reference.reachabilityFence(scope);
+        }
+    }
+
+    @ForceInline
+    public static
+    <V extends VectorSupport.Vector<E>, E>
+    void storeIntoMemorySegment(Class<? extends V> vmClass, Class<E> e, int length,
+                                V v,
+                                MemorySegmentProxy msp, long offset,
+                                VectorSupport.StoreVectorOperation<MemorySegmentProxy, V> defaultImpl) {
+        // @@@ Smarter alignment checking if accessing heap segment backing non-byte[] array
+        if (msp.maxAlignMask() > 1) {
+            throw new IllegalArgumentException();
+        }
+
+        try {
+            storeIntoMemorySegmentScoped(
+                    msp.scope(),
+                    vmClass, e, length,
+                    v,
+                    msp, offset,
+                    defaultImpl);
+        } catch (ScopedMemoryAccess.Scope.ScopedAccessError ex) {
+            throw new IllegalStateException("This segment is already closed");
+        }
+    }
+
+    @Scoped
+    @ForceInline
+    private static
+    <V extends VectorSupport.Vector<E>, E>
+    void storeIntoMemorySegmentScoped(ScopedMemoryAccess.Scope scope,
```

Comment:
I believe the rest of the class uses "internal" as a suffix for the scoped method

Reply:
Done.